src/wallet/wallet2.cpp

```diff
@@ -3824,14 +3824,30 @@ bool wallet2::verify_password(const std::string& keys_file_name, const epee::wip
 
 void wallet2::encrypt_keys(const crypto::chacha_key &key)
 {
-  m_account.encrypt_keys(key);
-  m_account.decrypt_viewkey(key);
+  if (m_watch_only)
+  {
+    m_account.encrypt_viewkey(key);
+    m_account.decrypt_viewkey(key);
```

Comment:
That's nop, it can be removed (same below).

Anyway, I think the check should be done in the ctor, here:

  if (!locked || w.is_unattended() || w.ask_password() != tools::wallet2::AskPasswordToDecrypt)

Reply:
I'm not sure I particularly agree, wouldn't it be better in the wallet, so the wallet can never go into an invalid state, not the unlocker's responsibility?

Reply:
Encrypting and decrypting the null secret key is no different from encrypting and decrypting an actual secret key, with regards to the wallet being in an invalid state. If the use your patch, the caller is still able to encrypt it twice, or decrypt it more than it was encrypted, etc. The only way to have it in a valid state is to have the keys encrypted either once, or zero times. And in those two cases, the state is valid whether the spend key is null_skey or not.

Maybe I'm missing a particular case where this makes a difference though ?

Reply:
No not in particular, just didn't understand the reasoning. But thinking at it in the way that decrypting the null key is a valid state is ok.

I've updated the commit, is that the desired fix?